packages/flutter/lib/src/rendering/object.dart

```diff
@@ -361,6 +364,282 @@ typedef void RenderingExceptionHandler(RenderObject source, String method, dynam
 /// information will be printed to the console instead.
 RenderingExceptionHandler debugRenderingExceptionHandler;
 
+class _SemanticsGeometry {
+  _SemanticsGeometry() : transform = new Matrix4.identity();
+  _SemanticsGeometry.withClipFrom(_SemanticsGeometry other) {
+    clipRect = other?.clipRect;
+    transform = new Matrix4.identity();
+  }
+  _SemanticsGeometry.copy(_SemanticsGeometry other) {
+    if (other != null) {
+      clipRect = other.clipRect;
+      transform = new Matrix4.copy(other.transform);
+    } else {
+      transform = new Matrix4.identity();
+    }
+  }
+  Rect clipRect;
+  Rect _intersectClipRect(Rect other) {
+    if (clipRect == null)
+      return other;
+    if (other == null)
+      return clipRect;
+    return clipRect.intersect(other);
+  }
+  Matrix4 transform;
+  void applyAncestorChain(List<RenderObject> ancestorChain) {
```

Comment:
In theory these's no redundant work here. For each parent-child step, we only ever compute it once. When there's multiple children, we fork the matrix (see the _ForkingSemanticsFragment class).